config/crd/bases/activemonitor.keikoproj.io_healthchecks.yaml

```diff
@@ -56,6 +56,48 @@ spec:
               type: string
             level:
               type: string
+            remedyworkflow:
```

Comment:
Is it possible that the remedy workflow is still in progress when the next healthcheck workflow starts? What's the default behavior right now? Do we need to add something in the CR to prevent such situations?

Reply:
No because the workflow jobs should be cleaned up with active-deadline seconds..
but will look into this more and test this if there are any corner cases

Reply:
It happens in this scenario given below:

```
apiVersion: activemonitor.keikoproj.io/v1alpha1
kind: HealthCheck
metadata:
  generateName: fail-healthcheck-
  namespace: health
spec:
  repeatAfterSec: 60 # duration in seconds
  level: cluster
  workflow:
    generateName: fail-workflow-
    resource:
      namespace: health # workflow will be submitted in this ns
      serviceAccount: activemonitor-controller-sa # workflow will be submitted using this
      source:
        inline: |
          apiVersion: argoproj.io/v1alpha1
          kind: Workflow
          spec:
            ttlSecondsAfterFinished: 60
            entrypoint: start
            templates:
            - name: start
              retryStrategy:
                limit: 3
              container:
                image: alpine
                command: [sh, -c]
                args: ["fail"]
  remedyworkflow:
    generateName: remedy-test-
    resource:
      namespace: health # workflow will be submitted in this ns
      serviceAccount: activemonitor-remedy-sa # workflow will be submitted using this acct
      source:
        inline: |
          apiVersion: argoproj.io/v1alpha1
          kind: Workflow
          spec:
            entrypoint: start
            templates:
            - name: start
              retryStrategy:
                limit: 3
              container:
                image: alpine
                command: [sh, -c]
                args: ["sleep 180"]
```

If we add ttlSecondsAfterFinished to the spec in remedy workflow.. I still see some pods are getting deleted but taking longer.
```
          spec:
            ttlSecondsAfterFinished: 60
```

Need to further debug and fix this case.

Reply:
The issues here is if we delete the healthcheck CR that in progress there is an issue in the current code as well that it will delete the CR but the workflow is not stopped and workflow-controller continues to run after the repeatAftersec interval.
I fixed it by checking once it comes out of the loop to see if the healthcheck Cr even exists or not.. if it doesn't exist return an error and stop processing the CR.

The other thing is activeDeadlineSeconds works as expected it sends SIGKill to the containers. The Pod then gets into an Error state if it is running some process and it gets deleted after ttlSecondsAfterFinished time.

Reply:
Please review this change and let me know if that is fine